### ISSUE TITLE
Getting dynamic learning rate of 'Adam'

### ISSUE THREAD

**Issue by bharatsau** [BVLC/caffe#4191]
Hi,

I was using solver type 'Adam' for my tests. As 'Adam' dynamically changes the learning rate, we dont need to train the net by using different learning rates. This reduces the headache as well as time to train a net.
However, if I had to resume training a net, in that case I generally get slightly worse result than the case where I train the net at one go without any pause for sufficient no of iterations.
I guess the reason may be: when resuming the training, 'base_lr' should be same as the learning rate when I paused the training. 
As we dont know the value of learning rate at each step while using 'Adam' solver type, it is difficult to set proper 'base_lr' when resuming the training.
I tried to check if there is any way to get the dynamic learning rate, but could not get any clue.

Kindly help if there is any way to get the dynamic learning rate. If there is no way to get it(other than changing the source code), may be this can be listed as a small enhancement of Caffe.

**Comment by seanbell** [comment]
ADAM uses a different effective learning rate for every parameter (and there are millions of parameters).  There's the base learning rate multiplier which can change every iteration depending on your settings, and for that I agree there should probably be a method added to Solver to extract the learning rate, which may or may not be well-defined for certain solver types -- but I would argue that it's not well-defined for ADAM.

Also, when you resume training, are you (a) restoring the solverstate as well as the caffemodel, and (b) are you making sure that the data layer is providing images in a random order or otherwise skips to where it was before?

I think this discussion belongs on the mailing list since it's more of a usage question, so I'm going to close it.  Please follow up on the mailing list.

From https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md:
> _Please do not post usage, installation, or modeling questions, or other requests for help to Issues._
Use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) instead. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe.

**Comment by bharatsau** [comment]
I tested this on Cifar10 dataset.

**Here is the common configuration:**
1. base_lr = 0.0001
2. batch_size= 64
3. Total no training samples = 100k
4. solver_type= Adam
5. Input Layer = 'HDF5Data' with 'shuffle'= true
6. random_seed= 1

Below are two cases:

**Case 1: Training directly 300k iterations:**

This is equivalent to (300k*64)/100k=192 epochs.
I got highest test accuracy = 80.46%

**Case 2: Two-step training:**

Trained 150k iterations first. Then took the **solverstate** and resumed it for another 150k iterations without changing base_lr.
So No of epochs in each stage= 150k*64/100k = 96 epochs.
The highest test accuracy = 80.20%

**Conclusion:**
Although the difference is not big, it is also not ignorable. I guess the base_lr plays a role here, this can be verified once we can get the proper base_lr. 
The random selection of samples in 2nd stage of 2nd case is probably different than the first case. This may also have some effect, however, I think it should not cause 0.26% difference.
